src/index/coinstatsindex.cpp

```diff
@@ -0,0 +1,407 @@
+                               __func__, iter_tip->GetBlockHash().ToString());
+            }
+
+            ReverseBlock(block, iter_tip);
+
+            iter_tip = iter_tip->GetAncestor(iter_tip->nHeight - 1);
+        } while (new_tip != iter_tip);
+    }
+
+    // During a reorg, we need to copy all hash digests for blocks that are getting disconnected from the
+    // height index to the hash index so we can still find them when the height index entries are
+    // overwritten.
+    if (!CopyHeightIndexToHashIndex(*db_it, batch, m_name, new_tip->nHeight, current_tip->nHeight)) {
+        return false;
+    }
+
+    if (!m_db->WriteBatch(batch)) return false;
+
+    return BaseIndex::Rewind(current_tip, new_tip);
+}
+
+static bool LookupOne(const CDBWrapper& db, const CBlockIndex* block_index, DBVal& result)
+{
+    // First check if the result is stored under the height index and the value there matches the
+    // block hash. This should be the case if the block is on the active chain.
+    std::pair<uint256, DBVal> read_out;
+    if (!db.Read(DBHeightKey(block_index->nHeight), read_out)) {
+        return false;
+    }
+    if (read_out.first == block_index->GetBlockHash()) {
+        result = std::move(read_out.second);
+        return true;
+    }
+
+    // If value at the height index corresponds to an different block, the result will be stored in
+    // the hash index.
+    return db.Read(DBHashKey(block_index->GetBlockHash()), result);
+}
+
+bool CoinStatsIndex::LookupStats(const CBlockIndex* block_index, CCoinsStats& coins_stats) const
+{
+    DBVal entry;
+    if (!LookupOne(*m_db, block_index, entry)) {
+        return false;
+    }
+
+    coins_stats.hashSerialized = entry.muhash;
+    coins_stats.nTransactionOutputs = entry.nTransactionOutputs;
+    coins_stats.nBogoSize = entry.nBogoSize;
+    coins_stats.nTotalAmount = entry.nTotalAmount;
+    coins_stats.nDiskSize = entry.nDiskSize;
+
+    return true;
+}
+
+uint256 CoinStatsIndex::currentHashInternal()
+{
+    unsigned char out[384];
+    m_muhash.Finalize(out);
+    return (TruncatedSHA512Writer(SER_DISK, 0) << out).GetHash();
+}
+
+// Reverse Block in case of reorg
+bool CoinStatsIndex::ReverseBlock(const CBlock& block, const CBlockIndex* pindex)
+{
+    CBlockUndo block_undo;
+    std::pair<uint256, DBVal> read_out;
+
+    if (pindex->nHeight > 0) {
+        if (!UndoReadFromDisk(block_undo, pindex)) {
+            return false;
+        }
+
+        if (!m_db->Read(DBHeightKey(pindex->nHeight - 1), read_out)) {
+            return false;
+        }
+
+        uint256 expected_block_hash = pindex->pprev->GetBlockHash();
+        if (read_out.first != expected_block_hash) {
+            return error("%s: previous block header belongs to unexpected block %s; expected %s",
+                         __func__, read_out.first.ToString(), expected_block_hash.ToString());
+        }
+    }
+
+    MuHash3072 do_muhash;
```

Comment:
Why is this called `do_muhash` rather than `undo_muhash`?

Reply:
I came up with do_muhash because I needed the contrary term to undo_muhash. Since we are talking about rolling back a block that has 'do' (utxo created in the block) and 'undo' data (utxos destroyed in the block). I was thinking if I should reverse these terms when a block is removed or keep the terms the same as they were when the block was added. For me it seemed less confusing to keep the naming the same as when the block was added. So 'do' data that was added when the block is added still called 'do' data when the block is rolled back, just that the 'do' data is removed from the utxo set this time. Either way I probably confusing so I will keep adding comments to make this more clear.

I should add that I only need these intermediate values because division is much slower than multiplication. Since these operations are cummutative I can multiply all the values that need to be removed from the muhash and then remove them all with just one division.